vspace/src/main/java/edu/asu/diging/vspace/core/services/impl/ExternalLinkManager.java

```diff
@@ -0,0 +1,123 @@
+import edu.asu.diging.vspace.core.services.ISpaceManager;
+
+@Transactional
+@Service
+public class ExternalLinkManager extends LinkManager implements IExternalLinkManager{
+
+    @Autowired
+    private ISpaceManager spaceManager;
+
+    @Autowired
+    private ExternalLinkRepository externalLinkRepo;
+
+    @Autowired
+    private ExternalLinkDisplayRepository externalLinkDisplayRepo;
+
+    @Autowired
+    private IExternalLinkFactory externalLinkFactory;
+
+    @Autowired
+    private IExternalLinkDisplayFactory externalLinkDisplayFactory;
+
+    @Override
+    public List<ILinkDisplay> getLinkDisplays(String spaceId) {
+        return new ArrayList<>(externalLinkDisplayRepo.findExternalLinkDisplaysForSpace(spaceId));
+    }
+
+    @Override
+    public void deleteLink(String linkId) {
+        Optional<ExternalLink> linkOptional = externalLinkRepo.findById(linkId);
+        if (!linkOptional.isPresent()) {
+            return;
+        }
+
+        ISpace space = linkOptional.get().getSpace();
+        IExternalLink link = linkOptional.get();
+        space.getExternalLinks().remove(link);
+        externalLinkDisplayRepo.deleteByExternalLink(link);
+        externalLinkRepo.delete((ExternalLink) link);
+
+    }
+
+    @Override
+    protected ILink createLinkObject(String title, String id, IVSpaceElement target, String linkLabel) {
+        ISpace source = spaceManager.getSpace(id);
+        IExternalLink link = externalLinkFactory.createExternalLink(title, source, ((ExternalLinkValue)target).getValue());
+        link.setName(linkLabel);
+        externalLinkRepo.save((ExternalLink) link);
+        return link;
+    }
+
+    @Override
+    protected IVSpaceElement getTarget(String externalLink) {
+        return new ExternalLinkValue(externalLink);
+    }
+
+    @Override
+    protected ILinkDisplay updateLinkAndDisplay(ILink link, ILinkDisplay displayLink) {
+        externalLinkRepo.save((ExternalLink) link);
+        externalLinkDisplayRepo.save((ExternalLinkDisplay) displayLink);
+        return displayLink;
+    }
+
+    @Override
+    protected void setTarget(ILink link, IVSpaceElement target) {
+        ((IExternalLink) link).setExternalLink(((ExternalLinkValue)target).getValue());
+
+    }
+
+    @Override
+    protected ILinkDisplay getDisplayLink(String externalLinkDisplayId) throws LinkDoesNotExistsException {
+        Optional<ExternalLinkDisplay> externalLinkOptional = externalLinkDisplayRepo.findById(externalLinkDisplayId);
+        linksValidation(externalLinkOptional);
+        IExternalLinkDisplay display = externalLinkOptional.get();
+        return display;
+    }
+
+    @Override
+    protected ILink getLink(String externalLinkIDValueEdit) throws LinkDoesNotExistsException {
+        Optional<ExternalLink> linkOptional = externalLinkRepo.findById(externalLinkIDValueEdit);
+        linksValidation(linkOptional);
+        IExternalLink link = linkOptional.get();
+        return link;
```

Comment:
this should return null if the link doesn't exist

Reply:
throws exception from linkValidation method if link is not present.

Reply:
then either the method shouldn't throw an exception (it can just return null which would be fine), or you need to handle the exception here. but all other `get...` methods return null if element is not found, the same should be true here.

Reply:
mrege with line 114